Update sncast scripts

franciszekjob opened the pull request:
<!-- Reference any GitHub issues resolved by this PR -->

Closes #3102
Towards #2490

## Introduced changes

<!-- A brief description of the changes -->

- Update `FeeSettings` struct
- Update and fix scripts tests

## Checklist

<!-- Make sure all of these are complete -->

- [x] Linked relevant issue
- [x] Updated relevant documentation
- [x] Added relevant tests
- [x] Performed self-review of the code
- [x] Added changes to `CHANGELOG.md`

franciszekjob committed:
Refactor `FeeSettings` in `sncast_std`

franciszekjob committed:
Merge branch 'franciszekjob/2490-rpc-0.8.0' of https://github.com/foundry-rs/starknet-foundry into franciszekjob/2490-2-fix-sncast-scripts

franciszekjob committed:
Add rpc changes to sncast_std

franciszekjob committed:
Update script tests

franciszekjob committed:
Merge branch 'franciszekjob/2490-rpc-0.8.0' of https://github.com/foundry-rs/starknet-foundry into franciszekjob/2490-2-fix-sncast-scripts

franciszekjob committed:
Fix some script tests

franciszekjob committed:
Merge branch 'franciszekjob/2490-rpc-0.8.0' of https://github.com/foundry-rs/starknet-foundry into franciszekjob/2490-2-fix-sncast-scripts

franciszekjob committed:
Fix linting

franciszekjob committed:
Add todos and fix tests

franciszekjob committed:
Merge branch 'franciszekjob/2490-rpc-0.8.0' of https://github.com/foundry-rs/starknet-foundry into franciszekjob/2490-2-fix-sncast-scripts

franciszekjob committed:
Merge branch 'franciszekjob/2490-rpc-0.8.0' of https://github.com/foundry-rs/starknet-foundry into franciszekjob/2490-2-fix-sncast-scripts

franciszekjob committed:
Merge branch 'franciszekjob/2490-rpc-0.8.0' of https://github.com/foundry-rs/starknet-foundry into franciszekjob/2490-2-fix-sncast-scripts

franciszekjob committed:
Merge branch 'franciszekjob/2490-rpc-0.8.0' of https://github.com/foundry-rs/starknet-foundry into franciszekjob/2490-2-fix-sncast-scripts

franciszekjob committed:
Merge branch 'franciszekjob/2490-rpc-0.8.0' of https://github.com/foundry-rs/starknet-foundry into franciszekjob/2490-2-fix-sncast-scripts

franciszekjob committed:
Merge branch 'franciszekjob/2490-rpc-0.8.0' of https://github.com/foundry-rs/starknet-foundry into franciszekjob/2490-2-fix-sncast-scripts

franciszekjob committed:
Add todos

franciszekjob committed:
Merge branch 'franciszekjob/2490-rpc-0.8.0' of https://github.com/foundry-rs/starknet-foundry into franciszekjob/2490-2-fix-sncast-scripts

franciszekjob committed:
Merge branch 'franciszekjob/2490-rpc-0.8.0' of https://github.com/foundry-rs/starknet-foundry into franciszekjob/2490-2-fix-sncast-scripts

franciszekjob committed:
Update changelog

franciszekjob committed:
Fix `test_wrong_calldata`

franciszekjob committed:
Fix `test_wrong_function_name`

franciszekjob committed:
Add todo

franciszekjob committed:
Merge branch 'franciszekjob/2490-rpc-0.8.0' of https://github.com/foundry-rs/starknet-foundry into franciszekjob/2490-2-fix-sncast-scripts

franciszekjob committed:
Merge branch 'franciszekjob/2490-rpc-0.8.0' of https://github.com/foundry-rs/starknet-foundry into franciszekjob/2490-2-fix-sncast-scripts

franciszekjob committed:
Merge branch 'franciszekjob/2490-rpc-0.8.0' of https://github.com/foundry-rs/starknet-foundry into franciszekjob/2490-2-fix-sncast-scripts

franciszekjob committed:
Remove docs snippets fix

franciszekjob committed:
Merge branch 'franciszekjob/2490-rpc-0.8.0' of https://github.com/foundry-rs/starknet-foundry into franciszekjob/2490-2-fix-sncast-scripts

franciszekjob committed:
Merge branch 'franciszekjob/2490-rpc-0.8.0' into franciszekjob/2490-2-fix-sncast-scripts

cptartur committed:
Fix cast script errors (#3099)

<!-- Reference any GitHub issues resolved by this PR -->

Related #3091

## Introduced changes

<!-- A brief description of the changes -->

- Resolved all issues with tests except for
`test_insufficient_account_balance`

## Checklist

<!-- Make sure all of these are complete -->

- [ ] Linked relevant issue
- [ ] Updated relevant documentation
- [ ] Added relevant tests
- [ ] Performed self-review of the code
- [ ] Added changes to `CHANGELOG.md`

---------

Co-authored-by: Franciszek Job <54181625+franciszekjob@users.noreply.github.com>

franciszekjob committed:
Merge branch 'franciszekjob/2490-rpc-0.8.0' of https://github.com/foundry-rs/starknet-foundry into franciszekjob/2490-2-fix-sncast-scripts

franciszekjob committed:
Merge branch 'franciszekjob/2490-2-fix-sncast-scripts' of https://github.com/foundry-rs/starknet-foundry into franciszekjob/2490-2-fix-sncast-scripts

franciszekjob committed:
Merge branch 'franciszekjob/2490-rpc-0.8.0' of https://github.com/foundry-rs/starknet-foundry into franciszekjob/2490-2-fix-sncast-scripts

ddoktorski commented on crates/sncast/tests/data/scripts/call/src/invalid_calldata.cairo:
```diff
@@ -2,19 +2,17 @@ use sncast_std::{call, CallResult, ScriptCommandError, ProviderError, StarknetEr
         array![0x12, 0x12, 0x12, 0x12, 0x12]
     )
-        .unwrap_err();
+        .unwrap();
```
Could you explain changes in this file?

ddoktorski replied:
Ok but this is related only to devnet?

franciszekjob replied:
Yes, it's a problem strictly with devnet.

ddoktorski commented on crates/sncast/tests/data/scripts/invoke/src/wrong_calldata.cairo:
```diff
@@ -8,15 +8,18 @@ fn main() {
     let map_contract_address = 0xcd8f9ab31324bb93251837e4efb4223ee195454f6304fcfcb277e277653008
         .try_into()
         .expect('Invalid contract address value');
+    let fee_settings = FeeSettings {
+        max_fee: Option::None,
+        l1_gas: Option::None,
+        l1_gas_price: Option::None,
+        l2_gas: Option::None,
+        l2_gas_price: Option::None,
+        l1_data_gas: Option::None,
+        l2_data_gas_price: Option::None,
+    };
```
Is it necessary to specify like this if everything is `None`? Maybe some helper function should be added to `sncast_std`?

franciszekjob replied:
Will be addressed in https://github.com/foundry-rs/starknet-foundry/issues/3102, hopefully will add it today and we can merge it before release.

franciszekjob replied:
I refactored it by adding 3 methods for `FeeSettings`: `max_fee`, `resource_bounds` and `estimate`. 
In case of `estimate`, it's a bit unintuitive because once the method is called, it doesn't fetch gas values (someone could assume so), but that's a remark.

ddoktorski commented on crates/sncast/tests/e2e/script/call.rs:
```diff
@@ -96,13 +97,15 @@ async fn test_call_invalid_calldata() {
-        ScriptCommandError::ProviderError(ProviderError::StarknetError(StarknetError::ContractError(ErrorData { msg: "Execution failed. Failure reason: 0x496e70757420746f6f206c6f6e6720666f7220617267756d656e7473 ('Input too long for arguments')." })))
-        ScriptCommandError::ProviderError(ProviderError::StarknetError(StarknetError::ContractError(ErrorData { msg: "Execution failed. Failure reason: 0x4661696c656420746f20646573657269616c697a6520706172616d202332 ('Failed to deserialize param #2')." })))
+        indoc! {r"
+        CallResult { data: [7733229381460288120802334208475838166080759535023995805565484692595] }
+        CallResult { data: [485748461484230571791265682659113160264223489397539653310998840191492914] }
```
Why is it printed this way now? Are we going to fix this?

franciszekjob replied:
It's a consequence of issue with call result coming as  from blockifier `Ok` instead of `Err`.
Linked todo for it.

franciszekjob replied:
> We should either link it to an issue to use devnet directly from crates/gh

What does it's change 😅 ?

ddoktorski commented on docs/listings/declare/src/lib.cairo:
```diff
@@ -1,16 +1,17 @@
 fn main() {
-    let max_fee = 9999999;
+    let fee_settings = FeeSettings {
+        max_fee: Option::Some(9999999),
```
What about a helper function that sets only `max_fee` and the rest fee settings to `None`?

franciszekjob replied:
Same as above.

franciszekjob replied:
Same as in https://github.com/foundry-rs/starknet-foundry/pull/3076#discussion_r2001154911

franciszekjob committed:
Merge branch 'franciszekjob/2490-rpc-0.8.0' of https://github.com/foundry-rs/starknet-foundry into franciszekjob/2490-2-fix-sncast-scripts

franciszekjob committed:
Add todo

franciszekjob commented on crates/sncast/tests/data/scripts/declare/src/with_invalid_max_fee.cairo:
```diff
@@ -3,20 +3,22 @@ use sncast_std::{
     assert(
         ScriptCommandError::ProviderError(
-            ProviderError::StarknetError(StarknetError::InsufficientMaxFee)
+            ProviderError::StarknetError(StarknetError::InsufficientResourcesForValidate)
```
Pleas see [errors in spec](https://github.com/starkware-libs/starknet-specs/blob/d70a2e57c9a66db1bbf86c388f38d6295a6a2a75/api/starknet_write_api.json#L228) - there isn't insufficient max fee error anymore.

franciszekjob committed:
Merge branch 'franciszekjob/2490-rpc-0.8.0' into franciszekjob/2490-2-fix-sncast-scripts

franciszekjob committed:
Merge branch 'franciszekjob/2490-rpc-0.8.0' of https://github.com/foundry-rs/starknet-foundry into franciszekjob/2490-2-fix-sncast-scripts

franciszekjob committed:
Merge branch 'franciszekjob/2490-rpc-0.8.0' of https://github.com/foundry-rs/starknet-foundry into franciszekjob/2490-2-fix-sncast-scripts

franciszekjob committed:
Merge branch 'franciszekjob/2490-2-fix-sncast-scripts' of https://github.com/foundry-rs/starknet-foundry into franciszekjob/2490-2-fix-sncast-scripts

franciszekjob committed:
Fix sncast docs snippets tests (#3101)

<!-- Reference any GitHub issues resolved by this PR -->

Closes #

## Introduced changes

<!-- A brief description of the changes -->


## Checklist

<!-- Make sure all of these are complete -->

- [x] Linked relevant issue
- [x] Updated relevant documentation
- [x] Added relevant tests
- [x] Performed self-review of the code
- [x] Added changes to `CHANGELOG.md`

franciszekjob committed:
Merge branch 'franciszekjob/2490-2-fix-sncast-scripts' of https://github.com/foundry-rs/starknet-foundry into franciszekjob/2490-2-fix-sncast-scripts

franciszekjob committed:
Add todos

franciszekjob committed:
Fix todos

franciszekjob committed:
Restore previous test content

franciszekjob committed:
Restore previous test content

franciszekjob committed:
Restore `invalid_calldata` script content

franciszekjob committed:
Fix asserted stdout content in `test_call_invalid_calldata` script test

franciszekjob committed:
Merge branch 'franciszekjob/2490-rpc-0.8.0' of https://github.com/foundry-rs/starknet-foundry into franciszekjob/2490-2-fix-sncast-scripts

franciszekjob committed:
Add `FeeSettingsImpl`; Adjust script tests

franciszekjob committed:
Add fee settings section in docs

franciszekjob committed:
Update changelog

cptartur commented on sncast_std/src/lib.cairo:
```diff
@@ -227,8 +229,58 @@ impl DisplayDeployResult of Display<DeployResult> {
 #[derive(Drop, Copy, Debug, Serde, PartialEq)]
 pub struct FeeSettings {
     pub max_fee: Option<felt252>,
-    pub max_gas: Option<u64>,
-    pub max_gas_unit_price: Option<u128>,
+    pub l1_gas: Option<u64>,
+    pub l1_gas_price: Option<u128>,
+    pub l2_gas: Option<u64>,
+    pub l2_gas_price: Option<u128>,
+    pub l1_data_gas: Option<u64>,
+    pub l2_data_gas_price: Option<u128>,
+}
```
I think we may want to start deprecating using these directly.

One idea, we could do something along the lines of this

```cairo
// TODO(#3102): Refactor to be enum with max fee and triplet variants
#[derive(Drop, Copy, Debug, Serde, PartialEq)]
#[deprecated(feature: "deprecated_fee_settings", note: "Use methods to create FeeSettings")]
// TODO rename feature and note
pub struct FeeSettings {
    pub max_fee: Option<felt252>,
    pub l1_gas: Option<u64>,
    pub l1_gas_price: Option<u128>,
    pub l2_gas: Option<u64>,
    pub l2_gas_price: Option<u128>,
    pub l1_data_gas: Option<u64>,
    pub l2_data_gas_price: Option<u128>,
}

// TODO rename
// New trait used internal for methods
pub struct FeeSettingsV2 {
    max_fee: Option<felt252>,
    l1_gas: Option<u64>,
    l1_gas_price: Option<u128>,
    l2_gas: Option<u64>,
    l2_gas_price: Option<u128>,
    l1_data_gas: Option<u64>,
    l2_data_gas_price: Option<u128>,
}

// TODO rename
trait Fees<T> {}

pub impl FeesFeeSetttingsV2Impl of Fees<FeeSettingsV2>;
pub impl FeeSettingsImpl of Fees<FeeSettings>;

// Change FeeSettingsImpl so it returns FeeSettingsV2 instead
```

and type our methods so their signature is along the lines of

```
pub fn invoke<T, +Fees<T>, +Serde<T>, +Drop<T>>(
    contract_address: ContractAddress,
    entry_point_selector: felt252,
    calldata: Array::<felt252>,
    fee_settings: T,
    nonce: Option<felt252>
) -> Result<InvokeResult, ScriptCommandError> {
```

This way no logic for these has to change and we can start pointing the users to using `FeeSettingsImpl` methods directly.

cptartur replied:
But I'm open to alternative methods of deprecating this, not sure if this is too complex or not.

cc @ddoktorski

franciszekjob replied:
As discussed offline: we can make fields in `FeeSettings` private in this PR, as we already make a breaking change.

ddoktorski replied:
Actually what's the rationale to make it private?

franciszekjob replied:
Mostly because we want to prevent from a situation when someone could pass `max_fee` and other gas values at the same time, which doesn't make sense.

ddoktorski replied:
Is it exactly the same logic as with the flags?

cptartur reviewed (commented):
Looks good, left some comments on how we can improve these

franciszekjob committed:
Merge branch 'franciszekjob/2490-rpc-0.8.0' of https://github.com/foundry-rs/starknet-foundry into franciszekjob/2490-2-fix-sncast-scripts

franciszekjob committed:
Remove `pub` from `FeeSettings` fields; Update docs

franciszekjob commented on crates/sncast/tests/e2e/script/declare.rs:
```diff
@@ -9,8 +9,9 @@ use test_case::test_case;
-#[test_case("braavos"; "braavos_account")]
+// TODO(#3089)
+// #[test_case("argent"; "argent_account")]
+// #[test_case("braavos"; "braavos_account")]
```
Yes, will forward it to them in a group message containing all our encountered issues.

ddoktorski replied:
Would it take a lot of effort to implement this within the scope of this PR?

franciszekjob commented on crates/sncast/tests/e2e/script/call.rs:
```diff
@@ -86,6 +86,7 @@ async fn test_call_invalid_address() {
 }
 
 #[tokio::test]
+#[ignore = "TODO(#3120)"]
```
Actually, it's because `execution_error` in `TransactionExecutionErrorData` (in sncast_std) should be now of type `ContractExecutionError` instead of `ByteArray`. Need to implement it.

franciszekjob replied:
Resolved.

ddoktorski replied:
Would it take a lot of effort to implement this within the scope of this PR?

franciszekjob replied:
Ok, so first, let me show how the implementation would ideally look:

```rust
#[derive(Drop, PartialEq, Serde, Debug, Copy)]
pub enum ContractExecutionError {
    Nested: Box<ContractExecutionErrorInner>,
    Message: ByteArray
}

#[derive(Drop, PartialEq, Serde, Debug, Copy)]
pub struct ContractExecutionErrorInner {
    contract_address: ContractAddress,
    class_hash: felt252,
    selector: felt252,
    error: ContractExecutionError,
}

```

The error is due to a recursive enum (`ContractExecutionError`) where one of the variants (`Nested`) holds a `Box<ContractExecutionErrorInner>`. To avoid infinite size, we box this variant (i.e., `Box<...>`).

For serialization, we require a custom serializer for both `ContractExecutionError` and `ContractExecutionErrorInner`. The serializer signature will look like:

```rust
fn serialize(self: @Box<ContractExecutionErrorInner>, ref output: Array<felt252>)
```

However, serialization with desnap requires the type to implement `Copy` (because non copyable type cannot be desnapped). So we attempt to derive `Copy` on `ContractExecutionError`.

Unfortunately, the `Message` variant contains a `ByteArray`, which itself is not `Copy` due to containing a `Vec<Felt>`. Therefore, we can't derive `Copy` on `ContractExecutionError` as-is.

To resolve this, the `Message` variant should be most likely wrapped in another type, but I need to think on the implementation. In rust there isn't such problem, but Cairo implementation is a bit more tricky.

franciszekjob replied:
Yes definitely, we're on the same page - just saying it can take a bit more time than expected.

franciszekjob committed:
Update todos

franciszekjob committed:
Fix linting

franciszekjob committed:
Remove changes from last two commits; Update todos

ddoktorski commented on sncast_std/src/lib.cairo:
```diff
@@ -223,12 +226,61 @@ impl DisplayDeployResult of Display<DeployResult> {
+    l2_gas: Option<u64>,
+    l2_gas_price: Option<u128>,
+    l1_data_gas: Option<u64>,
+    l2_data_gas_price: Option<u128>,
```
```suggestion
    l1_data_gas_price: Option<u128>,
```

ddoktorski commented on docs/src/appendix/sncast-library/declare.md:
```diff
@@ -46,14 +46,50 @@ pub struct DeclareTransactionResult {
+        l2_gas: u64,
+        l2_gas_price: u128,
+        l1_data_gas: u64,
+        l2_data_gas_price: u128
```
```suggestion
        l1_data_gas_price: u128
```

ddoktorski commented on docs/src/appendix/sncast-library/declare.md:
```diff
@@ -46,14 +46,50 @@ pub struct DeclareTransactionResult {
-    pub l1_data_gas_price: Option<u128>,
+
+#[generate_trait]
+pub impl FeeSettingsImpl of FeeSettingsTrait {
```
Can we add `FeeSettingsTrait` docs once and then reference it in relevant places instead of duplicating the same code multiple times?

ddoktorski commented on CHANGELOG.md:
```diff
@@ -32,6 +32,9 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 - `sncast completion` command - used to generate autocompletion script
 - `--l1-gas`, `--l1-gas-price`, `--l2-gas`, `--l2-gas-price`, `--l1-data-gas`, `--l1-data-gas-price` flags
+- `l1_gas`, `l1_gas_price`, `l2_gas`, `l2_gas_price`, `l1_data_gas`, `l1_data_gas_price` fields in `FeeSettings`
+- `FeeSettingImpl`, `FeeSettingsTrait`
```
This line seems redundant
```suggestion
```

ddoktorski replied:
I still think it is redundant since all the necessary information is included in the sentence below

franciszekjob committed:
Fix typos

franciszekjob committed:
Apply code review suggestions

franciszekjob committed:
Fix docs

ddoktorski commented on crates/sncast/tests/data/scripts/map_script/scripts/src/map_script.cairo:
```diff
@@ -44,17 +35,14 @@ fn second_contract() {
-        },
-        Option::Some(declare_nonce)
-    )
+    println!("DDD");
```
👀

ddoktorski commented on docs/src/starknet/script.md:
````diff
@@ -414,6 +411,35 @@ status: script panicked
+);
+```
+
+3. Specifying the maximum fee:
````
nit: Maybe it is worth mentioning that max fee is denoted in FRI?

franciszekjob committed:
Remove print

cptartur commented on CHANGELOG.md:
```diff
@@ -40,6 +42,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 #### Removed
 
 - `--max-gas` and `--max-gas-unit-price` flags
+- `max_gas`, `max_gas_unit_price` fields in `FeeSettings`
```
```suggestion
- `max_gas`, `max_gas_unit_price` fields in `FeeSettings` in `sncast_std`
```

cptartur commented on CHANGELOG.md:
```diff
@@ -32,6 +32,8 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 - `sncast completion` command - used to generate autocompletion script
 - `--l1-gas`, `--l1-gas-price`, `--l2-gas`, `--l2-gas-price`, `--l1-data-gas`, `--l1-data-gas-price` flags
+- `FeeSettingImpl`, `FeeSettingsTrait`
```
Please mention that these are in `sncast_std`.

cptartur commented on docs/src/appendix/sncast-library/fee_settings_trait.md:
```diff
@@ -0,0 +1,49 @@
+# `FeeSettingsTrait`
```
I think maybe just showing the signatures and explaining their intended usages would be better. Let's treat the returned struct as an implementation detail.

franciszekjob committed:
Apply code review suggestions

franciszekjob committed:
Update docs

franciszekjob committed:
Update changelog

franciszekjob committed:
Update changelog

ddoktorski commented on docs/src/appendix/sncast-library/fee_settings_trait.md:
````diff
@@ -0,0 +1,22 @@
+# `FeeSettingsTrait`
+
+```rust
+#[generate_trait]
+pub impl FeeSettingsImpl of FeeSettingsTrait {
````
nit:
```suggestion
pub trait FeeSettingsTrait {
```
In the current form it is not an implementation

franciszekjob replied:
Right, good catch.

franciszekjob committed:
Apply code review suggestion

franciszekjob committed:
Merge branch 'franciszekjob/2490-rpc-0.8.0' of https://github.com/foundry-rs/starknet-foundry into franciszekjob/2490-2-fix-sncast-scripts

franciszekjob committed:
Merge branch 'franciszekjob/2490-rpc-0.8.0' of https://github.com/foundry-rs/starknet-foundry into franciszekjob/2490-2-fix-sncast-scripts

franciszekjob committed:
Update changelog

franciszekjob committed:
Merge branch 'franciszekjob/2490-rpc-0.8.0' of https://github.com/foundry-rs/starknet-foundry into franciszekjob/2490-2-fix-sncast-scripts

cptartur commented on docs/src/starknet/script.md:
```diff
@@ -414,6 +411,35 @@ status: script panicked
+
+Passing fee settings is possible in a few different ways:
+
+1. Using auto estimation:
```
```suggestion
#### Using auto estimation:
```

cptartur replied:
Nit, but these multiline lists have weird formatting in my opinion

cptartur commented on docs/src/starknet/script.md:
````diff
@@ -414,6 +411,35 @@ status: script panicked
+let fee_settings = FeeSettingsTrait::estimate();
+```
+
+2. Manually setting the resource bounds (denoted in FRI):
````
```suggestion
#### Manually setting the resource bounds (denoted in FRI):
```

cptartur commented on docs/src/starknet/script.md:
````diff
@@ -414,6 +411,35 @@ status: script panicked
+);
+```
+
+3. Specifying the maximum fee (denoted in FRI):
````
```suggestion
#### Specifying the maximum fee (denoted in FRI):
```

franciszekjob committed:
Fix `FeeSettingsTrait` in sncast_std

franciszekjob committed:
Apply code review sugestion
